### PR TITLE
Cps/initialization of potential field

### PR DESCRIPTION
**Pull request by inigolcanalejo**
The user can now initialize the potential field. This is useful in the compressible case, because:

- not initializing it means that high local Mach numbers can appear in the first iteration of the solver near the inlet.
- it accelerates the convergence

This was also the idea and work for @marcnunezc 

Attention: this process will be soon migrated to c++

### CODE REVIEW

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py [hunk 1]
```diff
@@ -17,27 +16,30 @@ def __init__(self, Model, settings ):
...
+                "velocity_infinity": [1.0,0.0,0],
+                "initialize_flow_field": true
             }  """ );
 
 
         settings.ValidateAndAssignDefaults(default_parameters);
 
         self.model_part = Model[settings["model_part_name"].GetString()]
+        self.fluid_model_part = self.model_part.GetRootModelPart()
         self.velocity_infinity = KratosMultiphysics.Vector(3)#array('d', [1.0, 2.0, 3.14])#np.array([0,0,0])#np.zeros(3)#vector(3)
```
Comment (rubenzorrilla) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]:
```suggestion
        self.velocity_infinity = settings["velocity_infinity"].GetVector()
```

I think that this should work so you can avoid to seat one by one each one of the components.

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py [hunk 2]
```diff
@@ -17,27 +16,30 @@ def __init__(self, Model, settings ):
...
-        self.model_part.ProcessInfo.SetValue(CompressiblePotentialFlowApplication.VELOCITY_INFINITY,self.velocity_infinity)
+        self.inlet_potential = settings["inlet_phi"].GetDouble()
+        self.model_part.ProcessInfo.SetValue(CPFApp.VELOCITY_INFINITY,self.velocity_infinity)
+        self.initialize_flow_field = settings["initialize_flow_field"].GetBool()
 
 
 
     def Execute(self):
-        #KratosMultiphysics.VariableUtils().SetVectorVar(CompressiblePotentialFlowApplication.VELOCITY_INFINITY, self.velocity_infinity, self.model_part.Conditions)
+        #KratosMultiphysics.VariableUtils().SetVectorVar(CPFApp.VELOCITY_INFINITY, self.velocity_infinity, self.model_part.Conditions)
```
Comment (rubenzorrilla) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: Remove commented lines of code. You can always recover them by checking the history of the files.

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py [hunk 3]
```diff
@@ -59,6 +61,7 @@ def Execute(self):
```
Comment (rubenzorrilla) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: The comment says that this is done to find the minimal X. I feel that it is misleading...

**Review by rubenzorrilla — COMMENTED** [review]
Just for my understanding, would you mind explaining what you do in this process? From line 53 on it is not clear at all...

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py [hunk 4]
```diff
@@ -17,27 +16,30 @@ def __init__(self, Model, settings ):
                 "model_part_name":"PLEASE_CHOOSE_MODEL_PART_NAME",
                 "mesh_id": 0,
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]:
```suggestion

```
can safely be removed, is legacy

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py [hunk 5]
```diff
@@ -17,27 +16,30 @@ def __init__(self, Model, settings ):
...
+        self.initialize_flow_field = settings["initialize_flow_field"].GetBool()
 
 
 
     def Execute(self):
-        #KratosMultiphysics.VariableUtils().SetVectorVar(CompressiblePotentialFlowApplication.VELOCITY_INFINITY, self.velocity_infinity, self.model_part.Conditions)
+        #KratosMultiphysics.VariableUtils().SetVectorVar(CPFApp.VELOCITY_INFINITY, self.velocity_infinity, self.model_part.Conditions)
         for cond in self.model_part.Conditions:
-            cond.SetValue(CompressiblePotentialFlowApplication.VELOCITY_INFINITY, self.velocity_infinity)
+            cond.SetValue(CPFApp.VELOCITY_INFINITY, self.velocity_infinity)
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: why does every condition need this?

Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: https://github.com/KratosMultiphysics/Kratos/pull/4661#discussion_r275655518

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py (hunk 3 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: You are right, it is misleading. This process is cleaned up and better commented in https://github.com/KratosMultiphysics/Kratos/pull/4659. You can take a look in there and if there are still open questions I would gladly answer them. Also you can take a look at https://github.com/KratosMultiphysics/Kratos/pull/4661#issuecomment-483535939

**Comment by inigolcanalejo** [comment]
> Just for my understanding, would you mind explaining what you do in this process? From line 53 on it is not clear at all...

This process is cleaned up and better commented in https://github.com/KratosMultiphysics/Kratos/pull/4659. You can take a look in there and if there are still open questions I would gladly answer them. Also you can take a look at https://github.com/KratosMultiphysics/Kratos/pull/4661#issuecomment-483535939

**Comment by inigolcanalejo** [comment]
This is also ready to be merged

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py [hunk 6]
```diff
@@ -61,6 +64,12 @@ def Execute(self):
...
         self._AssignFarFieldBoundaryConditions(reference_inlet_node)
 
+        if(self.initialize_flow_field):
+            for node in self.fluid_model_part.Nodes:
+                initial_potential = DotProduct( node - reference_inlet_node, self.free_stream_velocity)
+                node.SetSolutionStepValue(CPFApp.VELOCITY_POTENTIAL,0,initial_potential + self.inlet_potential_0)
+                node.SetSolutionStepValue(CPFApp.AUXILIARY_VELOCITY_POTENTIAL,0,initial_potential + self.inlet_potential_0)
+
     def _FindFarthestUpstreamBoundaryNode(self):
         # The farthes upstream boundary node is the node with smallest
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: typo

**Review by marcnunezc — DISMISSED** [review]
I will approve considering that the comments by @philbucher and @rubenzorrilla have been implemented and tests are passing!
Minor comment on a typo.

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py (hunk 6 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: Corrected!

**Review by marcnunezc — APPROVED** [review]

**Comment by inigolcanalejo** [comment]
> I will approve considering that the comments by @philbucher and @rubenzorrilla have been implemented and tests are passing!

I rechecked and the comments by @philbucher and @rubenzorrilla were covered. However, the tests are giving segmentation fault at the moment.

**Comment by marcnunezc** [comment]
> > I will approve considering that the comments by @philbucher and @rubenzorrilla have been implemented and tests are passing!
> 
> I rechecked and the comments by @philbucher and @rubenzorrilla were covered. However, the tests are giving segmentation fault at the moment.

In my machine I get no errors! Do you have more info on the error you are getting?

**Comment by rubenzorrilla** [comment]
> > > I will approve considering that the comments by @philbucher and @rubenzorrilla have been implemented and tests are passing!
> > 
> > 
> > I rechecked and the comments by @philbucher and @rubenzorrilla were covered. However, the tests are giving segmentation fault at the moment.
> 
> In my machine I get no errors! Do you have more info on the error you are getting?

No errors in FullDebug mode or in Release mode?

**Comment by marcnunezc** [comment]
> > > > I will approve considering that the comments by @philbucher and @rubenzorrilla have been implemented and tests are passing!
> > > 
> > > 
> > > I rechecked and the comments by @philbucher and @rubenzorrilla were covered. However, the tests are giving segmentation fault at the moment.
> > 
> > 
> > In my machine I get no errors! Do you have more info on the error you are getting?
> 
> No errors in FullDebug mode or in Release mode?

I tested it in Release mode.

**Comment by rubenzorrilla** [comment]
> > > > > I will approve considering that the comments by @philbucher and @rubenzorrilla have been implemented and tests are passing!
> > > > 
> > > > 
> > > > I rechecked and the comments by @philbucher and @rubenzorrilla were covered. However, the tests are giving segmentation fault at the moment.
> > > 
> > > 
> > > In my machine I get no errors! Do you have more info on the error you are getting?
> > 
> > 
> > No errors in FullDebug mode or in Release mode?
> 
> I tested it in Release mode.

Can you test in FullDebug mode?

**Comment by marcnunezc** [comment]
> > > > > > I will approve considering that the comments by @philbucher and @rubenzorrilla have been implemented and tests are passing!
> > > > > 
> > > > > 
> > > > > I rechecked and the comments by @philbucher and @rubenzorrilla were covered. However, the tests are giving segmentation fault at the moment.
> > > > 
> > > > 
> > > > In my machine I get no errors! Do you have more info on the error you are getting?
> > > 
> > > 
> > > No errors in FullDebug mode or in Release mode?
> > 
> > 
> > I tested it in Release mode.
> 
> Can you test in FullDebug mode?

Yes! I tested it and it works as well in my machine, let's see if Iñigo can give us more info!

**Comment by inigolcanalejo** [comment]
Hi!

Sorry for my late reply. I must had had a problem with the compilation because I was getting segmentation fault on Monday. I pulled the branch again at it works so I will merge this.

**Pull request merged**